dali/operators/numba_function/numba_func.cc

```diff
@@ -0,0 +1,134 @@
+    types.CPointer(types.int32), # Output dtype pointer
+    types.CPointer(types.int64), # Input batch shape pointer
+    types.int32,                 # Number of dimensions in input shape
+    types.int32,                 # Input dtype
+    types.int32,                 # Number of samples in the batch
+    types.int32,                 # Number of outputs
+    types.int32                  # Number of inputs
+  )
+
+)code", 0);
+
+template <typename Backend>
+NumbaFunc<Backend>::NumbaFunc(const OpSpec &spec) : Base(spec) {
+  fn_ptr_ = spec.GetArgument<uint64_t>("fn_ptr");
+  setup_fn_ = spec.GetArgument<uint64_t>("setup_fn");
+}
+
+template<>
+void NumbaFunc<CPUBackend>::RunUserSetupFunc(std::vector<OutputDesc> &output_desc,
+    const workspace_t<CPUBackend> &ws) {
+  const auto &in = ws.InputRef<CPUBackend>(0);
+  if (!setup_fn_) {
+    output_desc[0] = {in.shape(), in.type()};
+    return;
+  }
+
+  auto in_shape = in.shape();
+  auto N = in_shape.num_samples();
+  auto ndim = in_shape.sample_dim();
+  TensorListShape<> output_shape;
+  output_shape.resize(N, ndim);
+  DALIDataType out_type = DALIDataType::DALI_NO_TYPE;
+  DALIDataType in_type = in.type().id();
+  ((void (*)(void*, int32_t, void*, const void*, int32_t, int32_t, int32_t))setup_fn_)(
+      output_shape.tensor_shape_span(0).data(), ndim, &out_type,
+      in_shape.tensor_shape_span(0).data(), ndim, in_type, N);
+
+  DALI_ENFORCE(out_type != DALIDataType::DALI_NO_TYPE,
+    "Output type was not set by the custom setup function.");
+  for (int i = 0; i < N; i++) {
+    for (int d = 0; d < ndim; d++) {
+      DALI_ENFORCE(output_shape.tensor_shape_span(i)[d] >= 0,
+        make_string(d, "-th", " dimension of ", i, "-th sample's shape is negative."));
+    }
+  }
+  output_desc[0].type = dali::TypeTable::GetTypeInfo(out_type);
+  output_desc[0].shape = output_shape;
+}
+
+template <>
+bool NumbaFunc<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+    const workspace_t<CPUBackend> &ws) {
+  const auto &in = ws.InputRef<CPUBackend>(0);
+  output_desc.resize(1);
+  RunUserSetupFunc(output_desc, ws);
+  return true;
+}
+
+template <>
+void NumbaFunc<CPUBackend>::RunImpl(workspace_t<CPUBackend> &ws) {
+  const auto &in = ws.InputRef<CPUBackend>(0);
+  auto in_shape = in.shape();
+  auto &out = ws.OutputRef<CPUBackend>(0);
+  auto out_shape = out.shape();
+  auto& tp = ws.GetThreadPool();
+
+  for (int sample_id = 0; sample_id < in_shape.num_samples(); sample_id++) {
+    tp.AddWork([&, fn_ptr = fn_ptr_, sample_id](int thread_id) {
+      ((void (*)(void*, const void*, int32_t, const void*, const void*, int32_t))fn_ptr)(
+        out[sample_id].raw_mutable_data(),
+        out_shape.tensor_shape_span(sample_id).data(),
+        out_shape.sample_dim(),
+        in[sample_id].raw_data(),
+        in_shape.tensor_shape_span(sample_id).data(),
+        in_shape.sample_dim());
+    }, out_shape.tensor_size(sample_id));
```

Comment:
I wonder if we should provide an ability to the user to give a hint regarding the workload each iteration may impose. But maybe it is too much.

Reply:
I think that's complicating things too much. Sorting by size is a good enough approximation in my opinion

Reply:
Just a thought.